pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/InStageStatsTreeBuilder.java

```diff
@@ -65,23 +66,82 @@ public InStageStatsTreeBuilder(MultiStageQueryStats.StageStats stageStats, IntFu
   }
 
   private ObjectNode selfNode(MultiStageOperator.Type type, Context context) {
+    return selfNode(type, context, _index, new JsonNode[0]);
+  }
+
+  private ObjectNode selfNode(MultiStageOperator.Type type, Context context, int index, JsonNode[] childrenArr) {
     ObjectNode json = JsonUtils.newObjectNode();
     json.put("type", type.toString());
-    for (Map.Entry<String, JsonNode> entry : _stageStats.getOperatorStats(_index).asJson().properties()) {
+    for (Map.Entry<String, JsonNode> entry : _stageStats.getOperatorStats(index).asJson().properties()) {
       json.set(entry.getKey(), entry.getValue());
     }
 
     if (json.get("parallelism") == null) {
       json.put("parallelism", context._parallelism);
     }
 
-    JsonNode executionTimeMs = json.get("executionTimeMs");
-    long cpuTimeMs = executionTimeMs == null ? 0 : executionTimeMs.asLong(0);
-    json.put("clockTimeMs", cpuTimeMs / context._parallelism);
+    addClockTimeMs(type, json, childrenArr, context);
+    addSelfAllocatedBytes(type, json, childrenArr, context);
+    addSelfGcTime(type, json, childrenArr, context);
+
+    if (childrenArr.length > 0) {
+      json.set(CHILDREN_KEY, JsonUtils.objectToJsonNode(childrenArr));
+    }
 
     return json;
   }
 
+  private void addClockTimeMs(MultiStageOperator.Type type, ObjectNode selfNode, JsonNode[] children, Context context) {
+    JsonNode executionTimeMs = selfNode.get("executionTimeMs");
+    long cpuTimeMs = executionTimeMs == null ? 0 : executionTimeMs.asLong(0);
+
+    selfNode.put("clockTimeMs", cpuTimeMs / context._parallelism);
+
+    long childrenStat = getChildrenStat(type, children, "executionTimeMs", false);
+    long selfExecutionTimeMs = cpuTimeMs - childrenStat;
+    if (selfExecutionTimeMs != 0) {
+      selfNode.put("selfExecutionTimeMs", selfExecutionTimeMs);
+      selfNode.put("selfClockTimeMs", selfExecutionTimeMs / context._parallelism);
+    }
+  }
+
+  private void addSelfAllocatedBytes(
+      MultiStageOperator.Type type, ObjectNode selfNode, JsonNode[] children, Context context) {
+    JsonNode allocatedBytes = selfNode.get("allocatedMemoryBytes");
+    long totalAllocatedBytes = allocatedBytes == null ? 0 : allocatedBytes.asLong(0);
+
+    long selfAllocatedBytes = totalAllocatedBytes - getChildrenStat(type, children, "allocatedMemoryBytes", false);
+    if (selfAllocatedBytes != 0) {
+      selfNode.put("selfAllocatedMB", selfAllocatedBytes / (1024 * 1024));
+    }
+  }
+
+  private void addSelfGcTime(MultiStageOperator.Type type, ObjectNode selfNode, JsonNode[] children, Context context) {
+    JsonNode gcTimeMs = selfNode.get("gcTimeMs");
+    long totalGcTimeMs = gcTimeMs == null ? 0 : gcTimeMs.asLong(0);
+
+    long selfGcTimeMs = totalGcTimeMs - getChildrenStat(type, children, "gcTimeMs", false);
+    if (selfGcTimeMs != 0) {
+      selfNode.put("selfGcTimeMs", selfGcTimeMs);
+    }
+  }
+
+  private long getChildrenStat(
```

Comment:
Which stats cross stage boundaries and which ones don't?

Reply:
None of the current stats that call this method cross boundaries. It is just here in case we want to support that in the future